### PR TITLE
fix(KL-181): handle cors error

### PR DESCRIPTION
**Pull request by ohhamma**
## 📌 연관된 이슈
- **[KL-181/cors 관련 문제 해결](https://ohhamma.atlassian.net/browse/KL-181)**

## 📝 작업 내용
- cors 관련 처리를 security로 해결
- 토큰 인증 및 발급을 쿠키로 하도록 수정
- oauth 로그인 성공 후 토큰과 함께 프론트엔드 엔드포인트로 리다이렉트하도록 처리
- 토큰이 만료된 후 인증이 필요없는 엔드포인트에 접속이 안되는 오류 해결

## 🌳 작업 브랜치명
- `KL-181/cors-관련-문제-해결`

## Release Notes

- **New Features**
	- Updated API endpoints for user image uploads and likes, focusing on user-specific interactions.
	- Introduced methods for retrieving user likes and following products.
	- Added functionality to update user information and retrieve member details through simplified endpoints.
	- Introduced utility classes for token management and pagination.

- **Bug Fixes**
	- Removed deprecated methods to streamline API and improve clarity.

- **Documentation**
	- Updated configuration files to reflect changes in application properties and security settings.

- **Tests**
	- Enhanced test coverage by introducing mock dependencies for new utilities and updating endpoint tests.

### CODE REVIEW

**Comment by ohhamma** [comment]
@coderabbitai review

**Review by idealflower-k — DISMISSED** [review]
수정사항 확인했습니다~! :)

File: src/main/java/taco/klkl/global/util/TokenUtil.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+public class TokenUtil {
+
+	public void addAccessTokenCookie(HttpServletResponse response, String accessToken) {
+		Cookie cookie = new Cookie(ACCESS_TOKEN, accessToken);
+		cookie.setHttpOnly(true);
+		cookie.setSecure(false); // TODO: HTTPS 사용 시 활성화
+		cookie.setPath("/");
+		cookie.setMaxAge(3600);
+		response.addCookie(cookie);
+	}
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/global/util/TokenUtil.java:20]: i'll add it later

File: src/main/java/taco/klkl/global/util/ProductUtil.java [hunk 2]
```diff
@@ -32,6 +32,10 @@ public Page<Product> findProductsByMemberId(final Long memberId, final Pageable
 		return productRepository.findByMemberId(memberId, pageable);
 	}
 
+	public Page<Product> findProductsByMemberIds(final Set<Long> memberIds, final Pageable pageable) {
+		return productRepository.findByMemberIds(memberIds, pageable);
+	}
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/global/util/ProductUtil.java]: it's outdated

File: src/test/java/taco/klkl/domain/member/controller/MemberControllerTest.java [hunk 3]
```diff
@@ -36,6 +37,9 @@ class MemberControllerTest {
 	@MockBean
 	private ResponseUtil responseUtil;
 
+	@MockBean
+	private TokenUtil tokenUtil;
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/test/java/taco/klkl/domain/member/controller/MemberControllerTest.java]: it's outdated

File: src/test/java/taco/klkl/domain/region/controller/currency/CurrencyControllerTest.java [hunk 4]
```diff
@@ -39,6 +40,9 @@ public class CurrencyControllerTest {
 	@MockBean
 	private ResponseUtil responseUtil;
 
+	@MockBean
+	private TokenUtil tokenUtil;
+
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/test/java/taco/klkl/domain/region/controller/currency/CurrencyControllerTest.java]: it's outdated

File: src/main/java/taco/klkl/domain/product/dao/ProductRepository.java [hunk 5]
```diff
@@ -9,21 +9,15 @@
...
-	@Query(
-		"SELECT p FROM product p WHERE p.member IN "
-			+ "(SELECT f.following FROM follow f WHERE f.follower = :follower "
-			+ "AND (:followingIds IS NULL OR f.following.id IN :followingIds))"
-	)
-	Page<Product> findProductsOfFollowedUsers(
-		@Param("follower") final Member follower,
-		@Param("followingIds") final Set<Long> followingIds,
-		final Pageable pageable
+	@Query("SELECT p FROM product p WHERE p.member.id IN :memberIds")
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/product/dao/ProductRepository.java]: my table name is `product`

File: src/test/java/taco/klkl/domain/region/controller/region/RegionControllerTest.java [hunk 6]
```diff
@@ -39,6 +40,9 @@ class RegionControllerTest {
 	@MockBean
 	private ResponseUtil responseUtil;
 
+	@MockBean
+	private TokenUtil tokenUtil;
+
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/test/java/taco/klkl/domain/region/controller/region/RegionControllerTest.java]: it's outdated

File: src/test/java/taco/klkl/domain/category/controller/category/CategoryControllerTest.java [hunk 7]
```diff
@@ -41,6 +42,9 @@ public class CategoryControllerTest {
 	@MockBean
 	private ResponseUtil responseUtil;
 
+	@MockBean
+	private TokenUtil tokenUtil;
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/test/java/taco/klkl/domain/category/controller/category/CategoryControllerTest.java]: it's outdated

File: src/test/java/taco/klkl/domain/like/controller/LikeControllerTest.java [hunk 8]
```diff
@@ -24,6 +24,7 @@
 import taco.klkl.global.config.security.TestSecurityConfig;
 import taco.klkl.global.error.exception.ErrorCode;
 import taco.klkl.global.util.ResponseUtil;
+import taco.klkl.global.util.TokenUtil;
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/test/java/taco/klkl/domain/like/controller/LikeControllerTest.java]: it's outdated

File: src/main/java/taco/klkl/domain/product/dao/ProductRepository.java [hunk 9]
```diff
@@ -9,21 +9,15 @@
...
+	Page<Product> findByMemberIds(
+		@Param("memberIds") Set<Long> memberIds,
+		Pageable pageable
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/product/dao/ProductRepository.java]: i'll change it later

File: src/main/java/taco/klkl/domain/like/controller/LikeController.java [hunk 10]
```diff
@@ -15,30 +17,35 @@
...
 public class LikeController {
 
 	private final LikeService likeService;
 
 	@GetMapping
-	@Operation(summary = "특정 상품의 좋아요 여부 조회", description = "특정 상품의 좋아요 여부를 조회합니다.")
-	public LikeResponse getLike(@PathVariable final Long productId) {
-		return likeService.getLike(productId);
+	@Operation(summary = "종아요 목록 조회", description = "좋아요를 누른 상품 목록을 조회합니다.")
+	public PagedResponse<ProductSimpleResponse> getMyLikes(
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/like/controller/LikeController.java]: i don't see any typo

File: src/main/java/taco/klkl/domain/like/controller/LikeController.java [hunk 11]
```diff
@@ -15,30 +17,35 @@
...
+		@PageableDefault(size = ProductConstants.DEFAULT_PAGE_SIZE) Pageable pageable
+	) {
+		return likeService.getLikes(pageable);
 	}
 
-	@PostMapping
+	@PostMapping("/products/{productId}")
 	@ResponseStatus(HttpStatus.CREATED)
 	@Operation(summary = "좋아요 누르기", description = "상품에 좋아요를 누릅니다.")
 	public LikeResponse addLike(@PathVariable final Long productId) {
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/like/controller/LikeController.java]: hmm i'll change it later

File: src/main/java/taco/klkl/domain/like/controller/LikeController.java [hunk 12]
```diff
@@ -15,30 +17,35 @@
...
 		return likeService.createLike(productId);
 	}
 
-	@DeleteMapping
+	@DeleteMapping("/products/{productId}")
 	@Operation(summary = "좋아요 취소", description = "상품에 누른 좋아요를 취소합니다.")
 	public LikeResponse removeLike(@PathVariable final Long productId) {
 		return likeService.deleteLike(productId);
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/like/controller/LikeController.java]: i'll think about it later

File: src/main/java/taco/klkl/domain/like/controller/LikeController.java [hunk 13]
```diff
@@ -15,30 +17,35 @@
...
 
 	@GetMapping
-	@Operation(summary = "특정 상품의 좋아요 여부 조회", description = "특정 상품의 좋아요 여부를 조회합니다.")
-	public LikeResponse getLike(@PathVariable final Long productId) {
-		return likeService.getLike(productId);
+	@Operation(summary = "종아요 목록 조회", description = "좋아요를 누른 상품 목록을 조회합니다.")
+	public PagedResponse<ProductSimpleResponse> getMyLikes(
+		@PageableDefault(size = ProductConstants.DEFAULT_PAGE_SIZE) Pageable pageable
+	) {
+		return likeService.getLikes(pageable);
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/like/controller/LikeController.java]: it was intended. i'll add another constant later

File: src/main/java/taco/klkl/global/config/security/TokenAuthenticationFilter.java [hunk 14]
```diff
@@ -68,10 +75,11 @@ private void setAuthentication(String accessToken) {
...
-		if (ObjectUtils.isEmpty(token) || !token.startsWith(TOKEN_PREFIX)) {
-			return null;
-		}
-		return token.substring(TOKEN_PREFIX.length());
+		return Optional.ofNullable(request.getCookies())
+			.flatMap(cookies -> Arrays.stream(cookies)
+				.filter(cookie -> ACCESS_TOKEN.equals(cookie.getName()))
+				.findFirst()
+				.map(Cookie::getValue))
+			.orElse(null);
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/global/config/security/TokenAuthenticationFilter.java]: i'll add it later

File: src/main/java/taco/klkl/global/config/security/TokenAuthenticationFilter.java [hunk 15]
```diff
@@ -46,7 +53,7 @@ protected void doFilterInternal(
 				String reissueAccessToken = tokenProvider.reissueAccessToken(accessToken);
 				if (StringUtils.hasText(reissueAccessToken)) {
 					setAuthentication(reissueAccessToken);
-					response.setHeader(AUTHORIZATION, TOKEN_PREFIX + reissueAccessToken);
+					tokenUtil.addAccessTokenCookie(response, reissueAccessToken);
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/global/config/security/TokenAuthenticationFilter.java]: i'll add it later

File: src/main/java/taco/klkl/global/config/security/SecurityConfig.java [hunk 16]
```diff
@@ -129,4 +109,26 @@ public SecurityFilterChain filterChain(HttpSecurity httpSecurity) throws Excepti
...
+
+	private RequestMatcher[] getUserRoleEndpoints() {
+		return SecurityEndpoint.USER_ROLE.getMatchers();
+	}
+
+	@Bean
+	public CorsConfigurationSource corsConfigurationSource() {
+		CorsConfiguration configuration = new CorsConfiguration();
+		configuration.setAllowedOrigins(List.of(applicationUri));
+		configuration.setAllowedMethods(List.of("GET", "POST", "PUT", "DELETE", "OPTIONS"));
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/global/config/security/SecurityConfig.java]: i'm currently using jdk17

File: src/main/java/taco/klkl/global/config/security/SecurityConfig.java [hunk 17]
```diff
@@ -35,6 +40,9 @@ public class SecurityConfig {
 	private final CustomAuthenticationEntryPoint authenticationEntryPoint;
 	private final TokenAuthenticationFilter tokenAuthenticationFilter;
 
+	@Value("${spring.application.uri}")
+	private String applicationUri;
+
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/global/config/security/SecurityConfig.java]: okay i'll change it later

File: src/main/java/taco/klkl/domain/member/service/MemberServiceImpl.java [hunk 18]
```diff
@@ -82,6 +73,20 @@ public List<MemberSimpleResponse> getFollowings() {
...
+	) {
+		final Pageable sortedPageable = pageableUtil.createPageableSortedByCreatedAtDesc(pageable);
+		final Set<Long> followingIds = Optional.ofNullable(memberIds)
+			.orElseGet(() -> getFollowings().stream()
+				.map(MemberSimpleResponse::id)
+				.collect(Collectors.toSet()));
+		final Page<Product> followingProducts = productUtil.findProductsByMemberIds(followingIds, sortedPageable);
+		return PagedResponse.of(followingProducts, ProductSimpleResponse::from);
+	}
+
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/member/service/MemberServiceImpl.java]: i'll handle it later

File: src/main/java/taco/klkl/domain/member/controller/MemberController.java [hunk 19]
```diff
@@ -63,52 +63,52 @@ public PagedResponse<ProductSimpleResponse> getMyProducts(
...
 
-	@PutMapping("/me")
-	@Operation(summary = "내 정보 수정", description = "내 정보를 수정합니다.")
-	public MemberDetailResponse updateMe(@Valid @RequestBody final MemberUpdateRequest request) {
-		return memberService.updateMember(request);
+	@GetMapping("/v1/me/following/products")
+	@Operation(summary = "내 팔로잉의 상품 목록 조회", description = "내 팔로잉 유저들의 상품 목록을 조회합니다.")
+	public PagedResponse<ProductSimpleResponse> getMyFollowingProducts(
+		@PageableDefault(size = ProductConstants.DEFAULT_PAGE_SIZE) final Pageable pageable,
+		@RequestParam(value = "member_id", required = false) final Set<Long> memberIds
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/member/controller/MemberController.java]: i intended it

File: src/main/java/taco/klkl/domain/product/dao/ProductRepository.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/product/dao/ProductRepository.java]: nope it's okay

File: src/main/java/taco/klkl/domain/like/controller/LikeController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/like/controller/LikeController.java]: okay i'll correct it later

**Review by idealflower-k — APPROVED** [review]
수정 확인했습니다~ :)

**Pull request merged**